src/spetlr/power_bi/PowerBi.py

```diff
@@ -0,0 +1,459 @@
+        if "access_token" not in result:
+            print(result)
+            self._raise_error("Failed to acquire token!")
+            return False
+
+        default_expires_in = 5 * 60
+        extra_seconds = 3
+
+        access_token = result["access_token"]
+        if "expires_in" in result:
+            self.expire_time = time.time() + result["expires_in"] - extra_seconds
+        else:
+            self.expire_time = time.time() + default_expires_in - extra_seconds
+
+        # Get latest Power BI Dataset refresh record
+        self.api_header = {
+            "Content-Type": "application/json",
+            "Authorization": f"Bearer {access_token}",
+        }
+        return True
+
+    @staticmethod
+    def _show_workspaces(df: pd.DataFrame):
+        print("Available workspaces:")
+        df.rename(
+            columns={"id": "workspace_id", "name": "workspace_name"},
+            inplace=True,
+        )
+        df.display()
+
+    @staticmethod
+    def _show_datasets(df: pd.DataFrame):
+        print("Available datasets:")
+        df.rename(
+            columns={"id": "dataset_id", "name": "dataset_name"},
+            inplace=True,
+        )
+        df.display()
+
+    def _get_workspace(self) -> bool:
+        """
+        Gets workspace ID based on the workspace name, or shows all workspaces
+            if no parameter was specified.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises SpetlrException: if failed and ignore_errors==False
+        """
+
+        if self.workspace_id is None:
+            api_call = requests.get(
+                url=f"{self.powerbi_url}groups", headers=self.api_header
+            )
+            if api_call.status_code != 200:
+                self._raise_api_error("Failed to fetch workspaces!", api_call)
+                return False
+            df = pd.DataFrame(api_call.json()["value"], columns=["id", "name"])
+
+            if self.workspace_name is None:
+                self._show_workspaces(df)
+                return False
+
+            rows = df.loc[df["name"] == self.workspace_name, "id"]
+            if rows.empty:
+                self._raise_error(
+                    f"Workspace name '{self.workspace_name}' cannot be found!"
+                )
+                return False
+            self.workspace_id = rows.values[0]
+
+        return True
+
+    def _get_dataset(self) -> bool:
+        """
+        Gets dataset ID based on the dataset name, or shows all datasets
+            if no parameter was specified.
+
+        :return: True if succeeded or False if failed (when ignore_errors==True)
+        :rtype: bool
+        :raises SpetlrException: if failed and ignore_errors==False
+        """
+
+        if self.dataset_id is None:
+            api_call = requests.get(
+                url=f"{self.powerbi_url}groups/{self.workspace_id}/datasets",
```

Comment:
FYI: Lets say the workspace_name is defined, but the  workspace_id is None, this api_call fails.